### ISSUE TITLE
Memory leak when using log tensor stats API from TransformerEngine

### ISSUE THREAD

**Issue by NVShreyas** [NVIDIA/nvidia-dlfw-inspect#4]
While integrating the NVDLFW Framework with Megatron-LM, we encountered an out-of-memory (OOM) issue when calling the log_tensor_stats API. The root cause is still unclear. However, we consistently observed the OOM only when logging activations during the forward pass — not when logging weights or gradients at training step end.
To isolate the problem, we tested using minimal changes on top of the original NVDLFW implementation from NVIDIA. Specifically, we invoked log_tensor_stats on a simple tensor (torch.tensor([0.0])) at the end of forward pass (after  line [392](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/models/gpt/gpt_model.py#L392)  in Megatron-LM). Even with this basic setup, the issue persisted during activation logging.

```yaml
nvdlfw_logging: # Config name. Used to differentiate between configs in the same file.
    enabled: True
    layers: # This field is used for layer selection.
        layer_name_regex_pattern: .*(embedding|attention|output|input|mlp|logits).* # 
    tensor_stats:
        LogTensorStats:
        enabled: True
        tensors_struct:
            - tensor: activation
            skip_reduction: false
            stats: [l2_norm]
            - tensor: weight
            skip_reduction: false
            stats: [l2_norm]
```

Assuming logging if off we still see memory leaks. However, if we toggle off logging and do an instant return without deepcopying in this parse_config_and_api function or then memory leaks disappear. In summary, if we remove the instant return and try either deepcopy or build a hardcoded “copy” both would leak. In addition, keeping this as an instant return without deepcopying then enabling logging function invocation brings memory leaks.
